File: app/src/main/java/org/dicio/dicio_android/output/speech/AndroidTtsSpeechDevice.java
package org.dicio.dicio_android.output.speech;

import android.content.Context;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.widget.Toast;

import androidx.annotation.NonNull;

import org.dicio.dicio_android.R;
import org.dicio.skill.output.SpeechOutputDevice;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class AndroidTtsSpeechDevice implements SpeechOutputDevice {

    private Context context;
    private TextToSpeech textToSpeech = null;
    private boolean initializedCorrectly = false;
    private final List<Runnable> runnablesWhenFinished = new ArrayList<>();
    private int lastUtteranceId = 0;

    public AndroidTtsSpeechDevice(final Context context, final Locale locale) {
        this.context = context;
        textToSpeech = new TextToSpeech(context, status -> {
            if (status == TextToSpeech.SUCCESS) {
                if (textToSpeech.setLanguage(locale) >= 0) { // errors are -1 or -2
                    initializedCorrectly = true;
                    textToSpeech.setOnUtteranceProgressListener(new UtteranceProgressListener() {
                        @Override
                        public void onStart(final String utteranceId) {
                        }

                        @Override
                        public void onDone(final String utteranceId) {
                            if (("dicio_" + lastUtteranceId).equals(utteranceId)) {
                                // run only when the last enqueued utterance is finished
                                for (final Runnable runnable : runnablesWhenFinished) {
                                    runnable.run();
                                }
                                runnablesWhenFinished.clear();
                            }
                        }

                        @Override
                        public void onError(final String utteranceId) {
                        }
                    });
                    return;
                } else {
                    Toast.makeText(context, R.string.android_tts_unsupported_language,
                            Toast.LENGTH_SHORT).show();
                }
            } else {
                Toast.makeText(context, R.string.android_tts_error, Toast.LENGTH_SHORT).show();
            }

            if (textToSpeech != null) {
                textToSpeech.shutdown();
                textToSpeech = null;
            }
        });
    }

    @Override
    public void speak(@NonNull final String speechOutput) {
        if (initializedCorrectly) {
            lastUtteranceId += 1;
            textToSpeech.speak(speechOutput, TextToSpeech.QUEUE_ADD, null,
                    "dicio_" + lastUtteranceId);
        } else {
            Toast.makeText(context, speechOutput, Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public void stopSpeaking() {
        if (textToSpeech != null) {
            textToSpeech.stop();
        }
    }

    @Override
    public boolean isSpeaking() {
        return textToSpeech != null && textToSpeech.isSpeaking();
    }

    @Override
    public void runWhenFinishedSpeaking(final Runnable runnable) {
        if (isSpeaking()) {
            runnablesWhenFinished.add(runnable);
        } else {
            runnable.run();
        }
    }

    @Override
    public void cleanup() {
        context = null;
        if (textToSpeech != null) {
            textToSpeech.shutdown();
            textToSpeech = null;
        }
    }
}

File: app/src/main/java/org/stypox/dicio/output/speech/AndroidTtsSpeechDevice.kt
package org.stypox.dicio.output.speech

import android.content.Context
import android.speech.tts.TextToSpeech
import android.speech.tts.TextToSpeech.OnInitListener
import android.speech.tts.UtteranceProgressListener
import android.widget.Toast
import androidx.annotation.StringRes
import org.dicio.skill.output.SpeechOutputDevice
import org.stypox.dicio.R
import java.util.Locale

class AndroidTtsSpeechDevice(private var context: Context, locale: Locale) : SpeechOutputDevice {
    private var textToSpeech: TextToSpeech? = null
    private var initializedCorrectly = false
    private val runnablesWhenFinished: MutableList<Runnable> = ArrayList()
    private var lastUtteranceId = 0

    init {
        textToSpeech = TextToSpeech(context, OnInitListener label@ { status: Int ->
            if (status == TextToSpeech.SUCCESS) {
                textToSpeech?.run {
                    if (setLanguage(locale) >= 0) { // errors are -1 or -2
                        initializedCorrectly = true
                        setOnUtteranceProgressListener(object :
                            UtteranceProgressListener() {
                            override fun onStart(utteranceId: String) {}
                            override fun onDone(utteranceId: String) {
                                if ("dicio_$lastUtteranceId" == utteranceId) {
                                    // run only when the last enqueued utterance is finished
                                    for (runnable in runnablesWhenFinished) {
                                        runnable.run()
                                    }
                                    runnablesWhenFinished.clear()
                                }
                            }

                            @Deprecated("")
                            override fun onError(utteranceId: String) {}
                        })
                    } else {
                        handleInitializationError(R.string.android_tts_unsupported_language)
                    }
                }
            } else {
                handleInitializationError(R.string.android_tts_error)
            }
        })
    }

    override fun speak(speechOutput: String) {
        if (initializedCorrectly) {
            lastUtteranceId += 1
            textToSpeech?.speak(
                speechOutput, TextToSpeech.QUEUE_ADD, null,
                "dicio_$lastUtteranceId"
            )
        } else {
            Toast.makeText(context, speechOutput, Toast.LENGTH_LONG).show()
        }
    }

    override fun stopSpeaking() {
        textToSpeech?.stop()
    }

    override fun isSpeaking(): Boolean {
        return textToSpeech?.isSpeaking == true
    }

    override fun runWhenFinishedSpeaking(runnable: Runnable) {
        if (isSpeaking) {
            runnablesWhenFinished.add(runnable)
        } else {
            runnable.run()
        }
    }

    override fun cleanup() {
        textToSpeech?.apply {
            shutdown()
            textToSpeech = null
        }
    }

    private fun handleInitializationError(@StringRes errorString: Int) {
        Toast.makeText(context, errorString, Toast.LENGTH_SHORT).show()
        cleanup()
    }
}